app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java

```diff
@@ -0,0 +1,573 @@
+                                        int id) throws IOException {
+        ContentValues updatedContentValues;
+        if (objectInDb) {
+            // was in db but is now to big, null db data entry and write to file
+            updatedContentValues = helper.getContentValuesWithCustomData(extObj, null);
+            updatedContentValues.put(DatabaseHelper.FILE_COL, newFilePath);
+            fileEncryptionKey = generateKeyAndAdd(updatedContentValues);
+        } else {
+            // was stored in a file all along, atomically update file by
+            // writing to new and removing old file
+            updatedContentValues = helper.getNonDataContentValues(extObj);
+            updatedContentValues.put(DatabaseHelper.FILE_COL, newFilePath);
+
+            HybridFileBackedSqlHelpers.setFileAsOrphan(db, currentFilePath);
+        }
+        db.update(table, updatedContentValues,
+                DatabaseHelper.ID_COL + "=?", new String[]{String.valueOf(id)});
+
+        writeStreamToFile(bos, newFilePath, fileEncryptionKey);
+        HybridFileBackedSqlHelpers.unsetFileAsOrphan(db, newFilePath);
+    }
+
+    @Override
+    public void remove(int id) {
+        SQLiteDatabase db = getDbOrThrow();
+
+        String filename = HybridFileBackedSqlHelpers.getEntryFilename(helper, table, id);
+        db.beginTransaction();
+        try {
+            db.delete(table, DatabaseHelper.ID_COL + "=?", new String[]{String.valueOf(id)});
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+
+        if (filename != null) {
+            File dataFile = new File(filename);
+            dataFile.delete();
+        }
+    }
+
+    @Override
+    public void remove(List<Integer> ids) {
+        if (ids.size() > 0) {
+            SQLiteDatabase db = getDbOrThrow();
+            db.beginTransaction();
+            List<String> filesToRemove;
+            try {
+                filesToRemove = HybridFileBackedSqlHelpers.getFilesToRemove(ids, helper, table);
+                List<Pair<String, String[]>> whereParamList = AndroidTableBuilder.sqlList(ids);
+                for (Pair<String, String[]> whereParams : whereParamList) {
+                    String whereClause = DatabaseHelper.ID_COL + " IN " + whereParams.first;
+                    db.delete(table, whereClause, whereParams.second);
+                }
+                db.setTransactionSuccessful();
+            } finally {
+                db.endTransaction();
+            }
+
+            HybridFileBackedSqlHelpers.removeFiles(filesToRemove);
+        }
+    }
+
+    @Override
+    public void removeAll() {
+        SQLiteDatabase db = getDbOrThrow();
+
+        db.beginTransaction();
+        try {
+            db.delete(table, null, null);
+            db.setTransactionSuccessful();
+        } finally {
+            db.endTransaction();
+        }
+
+        FileUtil.deleteFileOrDir(dbDir);
```

Comment:
But what if the file/dir deletion partially completes and then fails? Then the db entry would persist because the transaction rolled back, but the files would be missing